### PR TITLE
UPD: JuMP v1.15

### PR DESCRIPTION
**Pull request by pseudocubic**
Updates to support JuMP v1.15 nonlinear syntax.

@odow @ccoffrin We had to make some strange changes to the expressions and constraints in order to fix the failing tests from #454. In particular, if you look at something like any of the `constraint_mc_power_balance` functions, we had to put a negative sign inside the `sum` calls. By applying this to power or current variables we were able to get tests to pass. See for example line 151 in acp.jl or line 288 in objective.jl. 

Any ideas on why this works? We want to hold off on merging if we can until we have a grasp on the reason behind this.

CC @juanjospina

### CODE REVIEW

File: src/core/objective.jl [hunk 1]
```diff
@@ -277,24 +277,24 @@ function _objective_mc_min_fuel_cost_polynomial_linquad(pm::AbstractUnbalancedPo
...
+                #to avoid function calls inside of @constraint:
                 pg = var(pm, n, :pg, i)
                 pg = isa(pg, JuMP.Containers.DenseAxisArray) ? pg.data : pg
 
                 int_dim = length(pg)
                 if length(gen["cost"]) == 1
                     gen_cost[(n,i)] = gen["cost"][1]
                 elseif length(gen["cost"]) == 2
-                    gen_cost[(n,i)] = JuMP.@NLexpression(pm.model, gen["cost"][1]*sum(pg[i] for i in 1:int_dim) + gen["cost"][2])
+                    gen_cost[(n,i)] = JuMP.@expression(pm.model, (-gen["cost"][1]*sum(-pg[i] for i in 1:int_dim)) + gen["cost"][2])
```
Comment (odow) [inline src/core/objective.jl:288]: This should really not be needed. Let me take a look through the rest of the changes. Presumably there are now two signs that cancel each other out?

File: Project.toml [hunk 2]
```diff
@@ -29,7 +29,7 @@ Graphs = "1"
 InfrastructureModels = "0.7.3, 0.7.5"
 Ipopt = "0.9, 1.0.2, 1.1"
 JSON = "0.18, 0.19, 0.20, 0.21"
-JuMP = "0.22, 0.23, 1"
+JuMP = ">=1.15,<2"
```
Comment (odow) [inline Project.toml:32]:
```suggestion
JuMP = "1.15"
```
This should be equivalent.

File: src/form/acp.jl [hunk 3]
```diff
@@ -149,39 +144,39 @@ function constraint_mc_power_balance_slack(pm::AbstractUnbalancedACPModel, nw::I
...
+            + Pd[idx]
+            + ( # shunt
+                (Gt[idx,idx] * vm[t]^2)
+                +sum( (Gt[idx,jdx] * vm[t]*vm[u] * cos(va[t]-va[u]))
+                     +(Bt[idx,jdx] * vm[t]*vm[u] * sin(va[t]-va[u]))
                      for (jdx,u) in ungrounded_terminals if idx != jdx)
             )
-            + p_slack[t]
+            - p_slack[t]
+            == 0.0
```
Comment (odow) [inline src/form/acp.jl:161]: Why were these changes necessary?

File: src/form/acp.jl [hunk 4]
```diff
@@ -984,17 +979,17 @@ function constraint_mc_load_current_delta(pm::AbstractUnbalancedACPModel, nw::In
     vm_a, vm_b, vm_c = var(pm, nw, :vm, load_bus_id)
     va_a, va_b, va_c = var(pm, nw, :va, load_bus_id)
     # v_xy = v_x - v_y
-    vre_xy(vm_x, va_x, vm_y, va_y) = JuMP.@NLexpression(pm.model, vm_x*cos(va_x)-vm_y*cos(va_y))
-    vim_xy(vm_x, va_x, vm_y, va_y) = JuMP.@NLexpression(pm.model, vm_x*sin(va_x)-vm_y*sin(va_y))
+    vre_xy(vm_x, va_x, vm_y, va_y) = JuMP.@expression(pm.model, (vm_x*cos(va_x))-(vm_y*cos(va_y)))
+    vim_xy(vm_x, va_x, vm_y, va_y) = JuMP.@expression(pm.model, (vm_x*sin(va_x))-(vm_y*sin(va_y)))
```
Comment (odow) [inline src/form/acp.jl:983]: So really just a question: you've added all the extra brackets everywhere just to be safe? They really shouldn't be necessary.

File: src/core/objective.jl (hunk 1 above)
Comment (pseudocubic) [reply src/core/objective.jl:288]: It is very strange, we don't really understand why these are needed, but they fix a lot of the failing tests. The signs cancel out here, so this should be the exact same expression mathematically, but it affects the stability of the tests for some reason. Obviously, all of these tests pass with the old NLexpression syntax.

File: src/form/acp.jl (hunk 3 above)
Comment (pseudocubic) [reply src/form/acp.jl:161]: We found a pattern that fixed some tests, and applied it across all the power (current) balance equations. It is possible that moving all items to one side of the equal sign is not strictly necessary, and only changing the sign in `-sum(pg)` to `+sum(-pg)` is what was needed to get the tests to pass.

File: src/form/acp.jl (hunk 4 above)
Comment (pseudocubic) [reply src/form/acp.jl:983]: We agree that we don't think they should be necessary. This was just part of our debugging where we were trying to figure out if extra parentheses were needed or it was adding cancelling minus signs in some equations. I suspect the latter, we will have to selectively revert these changes to double check.

**Comment by pseudocubic** [comment]
I have reduced the changes to as few as I could to get the tests to resolve. Below is a summary of what was resolved by which changes.

## Objective expressions

I added minus signs in front of gen_cost and pg for the linear terms in the objectives in the following functions:

- `_objective_mc_min_fuel_cost_polynomial_linquad`
- `_objective_mc_min_fuel_cost_polynomial_nl`
- `_objective_mc_min_fuel_cost_polynomial_nl_switch`
- `objective_variable_pg_cost`

I guess only changes to `_objective_mc_min_fuel_cost_polynomial_nl` and `_objective_mc_min_fuel_cost_polynomial_linquad` are strictly needed, but that's because of a lack of complete test coverage for the other functions.

These changes fixed the following testsets (9 tests total):

- `2-bus diagonal ivr opf`
- `3-bus balanced ivr opf`
- `3-bus unbalanced ivr opf`
- `capcontrol_ivr`
- `test generator configuration`

## Power/Current balance constraints

I added minus signs in front of summations of generator power/current and inside the summation in the following functions:

- acp.jl
  - `constraint_mc_power_balance_slack`
  - `constraint_mc_power_balance_shed`
  - `constraint_mc_power_balance_shed`
  - `constraint_mc_power_balance_simple`
  - `constraint_mc_power_balance`
  - `constraint_mc_power_balance_capc`
- acr.jl
  - `constraint_mc_power_balance`
  - `constraint_mc_power_balance_capc`
  - `constraint_mc_power_balance_shed`
- ivr.jl
  - `constraint_mc_current_balance`
  - `constraint_mc_current_balance_capc`

Strictly speaking only changes to `constraint_mc_current_balance` and `constraint_mc_power_balance_capc` are probably needed to pass the tests, but that's probably because of lack of full coverage on the others.

These changes fixed the following testsets (4 tests total):

- `3w transformer ac pf center-tap`
- `3w transformer acp opf center-tap`
- `3w transformer acr opf center-tap`
- `3w transformer ivr opf center-tap`

## Load constraints

I added cancelling minus signs to expressions in the following functions:

- acp.jl
  - `constraint_mc_load_power_wye`
  - `constraint_mc_load_power_delta`
- acr.jl
  - `constraint_mc_load_power_wye`
  - `constraint_mc_load_power_wye`

These changes fixed the following testsets (10 tests total):

- `loadmodels connection variations`

**Comment by pseudocubic** [comment]
Note that I am not yet sure exactly the problem with the `branch power magnitude bound` failing tests in Julia 1.6

**Comment by odow** [comment]
Okay, let me take a much deeper look here. There's obviously something weird going on. We should definitely hold off merging until we understand what.

**Comment by odow** [comment]
Not yet. I'll try get to it this week.

**Comment by juanjospina** [comment]
@odow Following up to see if there are any updates on this issue?

**Comment by odow** [comment]
Taking a look now

**Comment by odow** [comment]
Okaaaay. This is a bug somewhere on our side.

```Julia
julia> using JuMP, Ipopt

julia> function main(flag::Bool)
           model = Model(Ipopt.Optimizer)
           set_silent(model)
           @variable(model, u[i = 1:3] == 0.3 * i)
           @variable(model, v[i = 1:3] == 0.5 * i)
           @variable(model, x[i = 1:3] == 0.7 * i)
           @variable(model, y[i = 1:3] == 1.1 * i)
           @expression(model, pg[i in 1:3], @force_nonlinear((x[i] * u[i]) + (y[i] * v[i])))
           @expression(model, pg2, 0.5 * sum(pg[i] for i in 1:3))
           @expression(model, pg3, -0.5 * sum(-pg[i] for i in 1:3))
           if flag
               @objective(model, Min, pg2)
           else
               @objective(model, Min, pg3)
           end
           optimize!(model)
           return objective_value(model)
       end
main (generic function with 1 method)

julia> main(true)
5.32

julia> main(false)
10.26
```
Assuming that `-(-x) == x`, then these should have the same objective values...

I'll keep digging.

**Comment by odow** [comment]
Okay, we're modifying existing expressions:
```Julia
julia> using JuMP

julia> model = Model();

julia> @variable(model, x[1:2]);

julia> y = [NonlinearExpr(:+, Any[xi]) for xi in x]
2-element Vector{NonlinearExpr}:
 +(x[1])
 +(x[2])

julia> a = @expression(model, sum(y[i] for i in 1:2))
x[1] + x[2]

julia> y
2-element Vector{NonlinearExpr}:
 x[1] + x[2]
 +(x[2])
```

**Comment by odow** [comment]
I've opened an issue https://github.com/jump-dev/JuMP.jl/issues/3825. I'll get this fixed ASAP since it is a pretty severe correctness bug.

**Comment by odow** [comment]
We can close this PR in favor of #454
```
  [8e850b90] libblastrampoline_jll v5.11.0+0
  [8e850ede] nghttp2_jll v1.52.0+1
  [3f19e933] p7zip_jll v17.4.0+2
        Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading.
Precompiling project...
  4 dependencies successfully precompiled in 42 seconds. 81 already precompiled.
     Testing Running tests...
[ Main | Info ] : running opendss parser tests
[ Main | Info ] : running misc data handling tests
[ Main | Info ] : running power flow (pf) tests
[ Main | Info ] : running branch-flow power flow (pf_bf) tests
[ Main | Info ] : running optimal power flow (opf) tests
[ Main | Info ] : running branch-flow optimal power flow (opf_bf) tests
[ Main | Info ] : running current-voltage optimal power flow (opf_iv) tests
[ Main | Info ] : running storage tests
[ Main | Info ] : running debug (pdf) tests
[ Main | Info ] : running multinetwork tests
[ Main | Info ] : running transformer tests
[ Main | Info ] : running capacitor control tests
[ Main | Info ] : running load models tests
[ Main | Info ] : running generator configuration tests
[ Main | Info ] : running matrix shunt tests
[ Main | Info ] : running minimum load delta (mld) tests
[ Main | Info ] : running data model creation and conversion tests
[ Main | Info ] : running explicit neutral opf bound tests
[ Main | Info ] : running explicit neutral power flow tests
[ Main | Info ] : running explicit neutral power flow tests with native julia power flow solver
Test Summary:           | Pass  Broken  Total     Time
PowerModelsDistribution | 1076       2   1078  4m58.3s
     Testing PowerModelsDistribution tests passed 

shell> git status
On branch od/update
Your branch is up to date with 'oscar/od/update'.

nothing to commit, working tree clean
```

The fix was https://github.com/jump-dev/JuMP.jl/pull/3826.

I'll tag a new JuMP version in https://github.com/jump-dev/JuMP.jl/pull/3827, and then I'll set that as the minimum version in the compat for this package.

**Comment by odow** [comment]
I'm sorry for not chasing this down earlier. I kinda assumed that the fault was somewhere in this package!

**Comment by pseudocubic** [comment]
Closed in favor of #454
